### DISCUSSION TITLE
Hardware acceleration in Linux (h265)

### DISCUSSION THREAD

**Discussion by Iuckyone** [mifi/lossless-cut#1963] [General]
Hello, I'm using **Arch Linux, X11** and trying to get LosslessCut to work properly with H265 videos (**I've got an Nvidia GPU (!)**).
I have mainly used it for cutting my OBS videos on Windows. Those videos are in H265Main codec and when I'm on Windows, everything works just fine. However, on Linux when I import the same videos I get a "File not natively supported" message. Turning off the hardware acceleration for H265 codec in settings results in the black screen.

After a bit of research about hardware acceleration in Chromium, I suppose that there is no way to get hardware acceleration to work with Nvidia GPU in Chromium on Linux right now. Use of VA-API, which, according to the [wiki](https://wiki.archlinux.org/title/Chromium#Hardware_video_acceleration), is technically supported by Chromium, isn't possible with nvidia proprietary drivers. Although there is a "guide" on the same wiki page that says that installation of additional libraries is required for VA-API to work, 

> "For proprietary NVIDIA support, installing [libva-vdpau-driver-chromium](https://aur.archlinux.org/packages/libva-vdpau-driver-chromium/)AUR[[broken link](https://wiki.archlinux.org/title/Help:Procedures#Fix_broken_package_links): package not found] or [libva-vdpau-driver-vp9-git](https://aur.archlinux.org/packages/libva-vdpau-driver-vp9-git/)AUR is required." 

first link here is broken and on the AUR page of the second one the pinned comment states:

> With the removal of --use-gl=desktop in Chromium 110, this extension is extremely brittle and effectively dead for many Chromium users. The new way of handling things in Chromium may further make any VDPAU-based solution impossible to support. If video acceleration is critical for you, the only solution might be to move to Firefox.

There is also an [open-source driver](https://github.com/elFarto/nvidia-vaapi-driver) for enabling hardware acceleration via VA-API. However, it is mainly focused on Firefox and support for Chromium isn't available ([issue](https://github.com/elFarto/nvidia-vaapi-driver/issues/5)). 

There is no info online related to this particular issue with LosslessCut and whether there are any workarounds, so if you have any information related to this, please share it here! Thanks!

**Comment by mifi** [comment]
unfortuately i don't have much knowledge here, but feel free to share any findings and eventually i could make a linux page with common tricks and tips.

**Comment by dankmolot** [comment]
Enabling hardware acceleration with an argument `--enable-features=AcceleratedVideoDecodeLinuxGL`  allows H265 videos to play natively!

https://wiki.archlinux.org/title/Chromium#Hardware_video_acceleration

Here is an example without and with the argument
```
$ losslesscut
```
<img width="1493" height="933" alt="image" src="https://github.com/user-attachments/assets/e9795784-c19a-4173-aebf-814ad29996c4" />


```
$ losslesscut --enable-features=AcceleratedVideoDecodeLinuxGL 
```
<img width="1495" height="931" alt="image" src="https://github.com/user-attachments/assets/06af70a5-5be5-491f-81cf-7aa6ee270595" />

Tested on LosslessCut 3.67.2 (Electron 38) [AUR](https://aur.archlinux.org/packages/losslesscut-bin), Arch Linux, KDE and AMD GPU

**Reply by dankmolot** [reply]
For Flatpak I had to replace:
```sh
run --branch=stable --arch=x86_64 --command=/app/bin/run.sh no.mifi.losslesscut
```
with
```sh
run --branch=stable --arch=x86_64 --command=/app/lossless-cut/losslesscut no.mifi.losslesscut --no-sandbox --enable-features=AcceleratedVideoDecodeLinuxGL
```

since `run.sh` does not forward arguments, also it has `--no-sandbox` argument inside, that's why it was added too

Also created an issue, so the flag could be added directly https://github.com/flathub/no.mifi.losslesscut/issues/91

**Reply by pricklythorn** [reply]
For me, adding `--enable-features=AcceleratedVideoDecodeLinuxGL` does not fix the issue, FFmpeg-assisted playback disappears, but there is no image where the video is, seeking and audio playback still work. I'm on CachyOS 7.0.12-1, Plasma 6.7.0-1.1, NVIDIA 610.43.02-3 (4070 Mobile with hybrid graphics), and LosslessCut 3.69.0-1.

**Reply by dankmolot** [reply]
@pricklythorn  Could you send an output of a command `vainfo`?

**Reply by pricklythorn** [reply]
```
Trying display: wayland
vainfo: VA-API version: 1.23 (libva 2.23.0)
vainfo: Driver version: VA-API NVDEC driver [direct backend]
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            : VAEntrypointVLD
      VAProfileMPEG2Main              : VAEntrypointVLD
      VAProfileVC1Simple              : VAEntrypointVLD
      VAProfileVC1Main                : VAEntrypointVLD
      VAProfileVC1Advanced            : VAEntrypointVLD
      VAProfileH264Main               : VAEntrypointVLD
      VAProfileH264High               : VAEntrypointVLD
      VAProfileH264ConstrainedBaseline: VAEntrypointVLD
      VAProfileJPEGBaseline           : VAEntrypointVLD
      VAProfileHEVCMain               : VAEntrypointVLD
      VAProfileVP8Version0_3          : VAEntrypointVLD
      VAProfileVP9Profile0            : VAEntrypointVLD
      VAProfileAV1Profile0            : VAEntrypointVLD
      VAProfileHEVCMain10             : VAEntrypointVLD
      VAProfileHEVCMain12             : VAEntrypointVLD
      VAProfileVP9Profile2            : VAEntrypointVLD
      VAProfileHEVCMain444            : VAEntrypointVLD
      VAProfileHEVCMain444_10         : VAEntrypointVLD
      VAProfileHEVCMain444_12         : VAEntrypointVLD
```

**Reply by dankmolot** [reply]
Looks normal, maybe something could be found in LosslessCut logs? Does it show any errors when you launch it from a terminal and try to open the file? Also how the file is encoded, what codec?

**Reply by pricklythorn** [reply]
File is a CQP 17, yuv444p, NVENC HEVC main10 MKV.
From the terminal I get this error:
```
[10413:0621/130114.642092:ERROR:ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc:252] '--ozone-platform=wayland' is not compatible with Vulkan. Consider switching to '--ozone-platform=x11' or disabling Vulkan
```
But that happens regardless of if `--enable-features=AcceleratedVideoDecodeLinuxGL` is specified, I have also tried X11, same issue.
In LosslessCut's console output, the only notable thing is:
```
ffmpeg.ts:317 file-type returned unknown format Array(2) video/matroska
determineSourceFileFormat @ ffmpeg.ts:317
```

**Reply by dankmolot** [reply]
@pricklythorn try to remux/record videos as mp4. If you are using OBS, then use "Hybrid MP4".
Chrome does not support mkv files by default, maybe remuxing to mp4 will help? But if not then I recommend creating an issue. I could find existing issue, only https://github.com/mifi/lossless-cut/issues/2364#issuecomment-2768918911 which does not seem to fit here

**Reply by pricklythorn** [reply]
I tried remuxing to MP4 in FFmpeg, same issue. I'll probably create an issue about it, might be something to do with NVIDIA.
